imap_processing/swe/l2/swe_l2.py

```diff
@@ -195,6 +201,119 @@ def calculate_flux(l1b_dataset: xr.Dataset) -> npt.NDArray:
     return flux
 
 
+def put_data_into_angle_bins(
+    data: np.ndarray, angle_bin_indices: npt.NDArray[np.int_]
+) -> npt.NDArray:
+    """
+    Put data in its angle bins.
+
+    This function bins SWE data into 30 predefined angle bins
+    while preserving the original energy step structure. For each
+    full cycle, it assigns data to the corresponding angle bin
+    based on the provided indices.
+
+    Since multiple data points may fall into the same angle bin,
+    the function accumulates values and computes the average across
+    all 7 CEMs, ensuring that each bin contains a representative
+    mean value while maintaining the 7 CEM structure.
+
+    Parameters
+    ----------
+    data : numpy.ndarray
+        Data to put in bins. Shape: (full_cycle_data, energy_step, angle_bin, 7).
+    angle_bin_indices : numpy.ndarray
+        Indices of angle bins to put data in. Shape:
+        (full_cycle_data, energy_step, angle_bin).
+
+    Returns
+    -------
+    numpy.ndarray
+        Data in bins. Shape: (full_cycle_data, 24, 30, 7).
+    """
+    # Initialize with zeros instead of NaN because np.add.at() does not
+    # work with nan values. It results in nan + value = nan
+    binned_data = np.zeros((data.shape[0], 24, 30, 7), dtype=np.float64)
+
+    time_indices = np.arange(data.shape[0])[:, None, None]
+    energy_indices = np.arange(24)[None, :, None]
+
+    # Use np.add.at() to accumulate values into bins
+    np.add.at(binned_data, (time_indices, energy_indices, angle_bin_indices), data)
+
+    # Count occurrences in each bin to compute the mean.
+    # Ensure float dtype for division
+    bin_counts = np.zeros_like(binned_data, dtype=float)
+    np.add.at(bin_counts, (time_indices, energy_indices, angle_bin_indices), 1)
+
+    # Compute the mean. Replace zero counts with NaN to indicate no data in the bin
+    # because zero physical counts could be valid data.
+    bin_counts[bin_counts == 0] = np.nan
+    binned_data /= bin_counts
+
+    return binned_data
+
+
+def find_angle_bin_indices(
+    inst_spin_angle: np.ndarray,
+    spin_angle_bin_edges: np.ndarray,
```

Comment:
I think these are fixed based on your comment below. Can we remove this argument so you are just looking up in your fixed lookup table essentially? If we do need to change it later, perhaps add the argument then.